components/script/dom/bindings/str.rs

```diff
@@ -731,52 +468,67 @@ fn parse_date_component(value: &str) -> Option<(i32, u32, u32)> {
     let hour = iterator.next()?;
     if hour.len() != 2 {
         return None;
     }
-    let hour_int = hour.parse::<u32>().ok()?;
-
-    // Step 2
+    // Step 2: If hour is not a number in the range 0 ≤ hour ≤ 23, then fail.
+    let hour_int = hour.parse::<u8>().ok()?;
     if hour_int > 23 {
         return None;
     }
 
-    // Step 3, 4
+    // Step 3: If position is beyond the end of input or if the character at position is
+    // not a U+003A COLON character, then fail. Otherwise, move position forwards one
+    // character.
+    // Step 4: Collect a sequence of code points that are ASCII digits from input given
+    // position. If the collected sequence is not exactly two characters long, then fail.
+    // Otherwise, interpret the resulting sequence as a base-ten integer. Let that number
+    // be the minute.
+    // Step 5: If minute is not a number in the range 0 ≤ minute ≤ 59, then fail.
     let minute = iterator.next()?;
     if minute.len() != 2 {
         return None;
     }
-    let minute_int = minute.parse::<u32>().ok()?;
-
-    // Step 5
+    let minute_int = minute.parse::<u8>().ok()?;
     if minute_int > 59 {
         return None;
     }
 
-    // Step 6, 7
-    let second_float = match iterator.next() {
-        Some(second) => {
-            let mut second_iterator = second.split('.');
-            if second_iterator.next()?.len() != 2 {
-                return None;
-            }
-            if let Some(second_last) = second_iterator.next() {
-                if second_last.len() > 3 {
-                    return None;
-                }
-            }
+    // Step 6, 7: Asks us to parse the seconds as a floating point number, but below this
+    // is done as integral parts in order to avoid floating point precision issues.
+    let Some(seconds_and_milliseconds) = iterator.next() else {
+        return Some((hour_int, minute_int, 0, 0));
+    };
 
-            second.parse::<f64>().ok()?
-        },
-        None => 0.0,
+    // Parse the seconds portion.
+    let mut second_iterator = seconds_and_milliseconds.split('.');
+    let second = second_iterator.next()?;
+    if second.len() != 2 {
+        return None;
+    }
+    let second_int = second.parse::<u8>().ok()?;
+
+    // Parse the milliseconds portion as a u16 (milliseconds can be up to 1000) and
+    // make sure that it has the proper value based on how long the string is.
+    let Some(millisecond) = second_iterator.next() else {
+        return Some((hour_int, minute_int, second_int, 0));
     };
+    let millisecond_length = millisecond.len() as u32;
+    if millisecond_length > 3 {
+        return None;
+    }
+    let millisecond_int = millisecond.parse::<u16>().ok()?;
+    let millisecond_int = millisecond_int * 10_u16.pow(3 - millisecond_length);
 
-    // Step 8
-    Some((hour_int, minute_int, second_float))
+    // Step 8: Return hour, minute, and second.
+    Some((hour_int, minute_int, second_int, millisecond_int))
```

Comment:
and, in our case, millis?

Reply:
Good point. I've updated this comment to reflect that in our case we return milliseconds.